modules/core/src/main/java/org/apache/ignite/cache/affinity/Affinity.java

```diff
@@ -188,8 +192,20 @@
      *      always first.
      * @throws IgniteException If there are no alive nodes for this cache.
      */
+    @Deprecated
     public Collection<ClusterNode> mapKeyToPrimaryAndBackups(K key);
 
+    /**
+     * Gets primary and backup nodes for the key. Primary node is always
+     * first in the returned list.
+     *
+     * @param key Key to get affinity nodes for.
+     * @return List of primary and backup nodes for the key with primary node
+     *      always first.
+     * @throws IgniteException If there are no alive nodes for this cache.
+     */
+    public List<ClusterNode> mapKeyToPrimaryAndBackupsList(K key);
```

Comment:
We should have default implementation of this method to preserve backward compatibility.
Something like:
```
Collection<ClusterNode> res = mapKeyToPrimaryAndBackups(key);
if (res instanceof List)
    return (List<ClusterNode>)res;
else
    return new ArrrayList(res);
```

Reply:
Can you please explain why do we need it? Clients with old version would use old method which hadn't been changed. All interface implementations are cited in the same module as I see

Reply:
Classes located in `org/apache/ignite/internal` package are Internal API.
`org/apache/ignite/cache/affinity/Affinity.java` package are Public API.
So, we should keep Public API backwards compatible between minor releases.
It means, we can't add new methods to the public API without default implementation, because, there are implementations of this interface out of Ignite codebase.